### ISSUE TITLE
How to avoid CPU spikes in step function

### ISSUE THREAD

**Issue by squinkylabs** [VCVRack/Tutorial#9]
I’m wondering if the VCVRack does anything to avoid CPU usage spikes due to floating point underflow and denormals in the module’s step() function.

I have seen DSP host software that always sends a signal (at -190 db) to try to keep poorly designed plugins from doing this.

Of course the “best practice” is to use SSE/AVX, and have the host be sure that the audio render thread is set up so that SSE denormals and underflows are just converted to zero with no CPU hit.
So – does VCVRack do either of these things?

**Comment by AndrewBelt** [comment]
Yes, it attempts to disable denormals in https://github.com/VCVRack/Rack/blob/v0.5/src/engine.cpp#L118. Another requirement is to compile with at least SSE enabled, which is done for both Rack and plugins in https://github.com/VCVRack/Rack/blob/v0.5/compile.mk#L9

Let me know if you experience denormals in practice for some architecture.

**Comment by squinkylabs** [comment]
Ah, I see. So those those mysterious compiler flags force the compiler to use SSE versions of scalar math operation? Nice. I must study -ffast-math -fno-finite-math-only

In the rack itself, shouldn't you also do this? _MM_SET_DENORMALS_ZERO_MODE(_MM_DENORMALS_ZERO_ON) https://software.intel.com/en-us/node/682949

**Comment by AndrewBelt** [comment]
Thanks, added in f31c132 of Rack repo.
I have confirmed that I don't get CPU spikes in all 5 environments I test on, but this probably won't hurt to add.

**Comment by squinkylabs** [comment]
thanks.

I don't know if you have a stripped down "test host" for modules, kind of like the VST host that comes with the VST SDK? If so you could have it trap x87 floating point exceptions, and pump the modules with various setting.

I don't remember, but I think I used to use the VST host to CPU profile my plugins - don't think I did anything about FP, since I explicitly used SSE intrinsics there.

**Comment by AndrewBelt** [comment]
Nope, that would be virtually impossible because plugins can access all header files of Rack and binds to it like an octopus.

Rack will always be super lightweight, so there's no need for a lightweight test host. Just test in the main application. Rack is open-source, unlike most VST hosts, so you can add traps to the `engineStep()` function as you wish.

**Comment by squinkylabs** [comment]
I think what you are saying is that individual developers can figure out how to profile their code, if they are sufficiently motivated. But out in the real world, how are users able to know which modules have ok real-time performance, vs ones that have horrible spikes that make them almost unusable?

Thinking about DAW plugins, they have this problem, too. AFAIK they address it in two ways: a) DAW developers themselves thoroughly test their own plugins to ensure that they are well behaved. b) Reviews and word of mouth let people know that some developers have nice stable plugins, and others don't.

I don't see any clear way to keep VSVRack modules "honest", and I don't yet see any rumors among users about which modules are prone popping and which aren't. In fact I haven't yet seen much acknowledgement that popping is in fact most often due to poor modules rather than VCVRack itself.

Ok, end of tirade... I guess I should look into making a tool to help...

**Comment by AndrewBelt** [comment]
I'll re-add the "CPU time monitor bar" in v1.0 (or maybe even v0.6), but the problem is that users will misuse the information. It's almost inevitable, but it can be reduced through clever hiding of the function.

**Comment by squinkylabs** [comment]
Sounds like a great idea. And yes, people will misuse it. Does this bar have peak and average display? That could be cool.

**Comment by AndrewBelt** [comment]
It's just an exponentially smoothed moving average. The noise for measuring time on the 1us scale is too large for a meaningful peak measurement.

**Comment by squinkylabs** [comment]
Good point. Of course a spike in a single step() isn't so bad anyway. I'm guessing you miss the hardware deadline if the total time for all the calls to step() for a given hardware buffer gets longer than the time to play that buffer? So the interesting thing would be the peak in the sum of all the steps() per "buffer". Of course you may do your scheduling in some other way that makes this more difficult. But that is the number you want: "how close am I to generating a pop (underflow)?", right? So you want a "peak" meter of the total time processing the last buffer.

**Comment by AndrewBelt** [comment]
How big are your plugins' buffers? The highest I've ever gone is 48, which is still under the minimum audio buffer size.

**Comment by squinkylabs** [comment]
Oh, I haven't written it yet, but it will use a standard 1 sample (no buffer) buffer. I meant the hardware's buffer size, or more accurately the VCVRack engine's buffer size (again, assuming that you will pop if the sum of all the steps for a given buffer is too long). Although I can easily envision that I might update "control voltages" at less that the audio sample rate (which might give me an effective buffer size of 16). I'm not worrying too much about the scheduling at this point. It's just my nature to obsess about those kinds of thing and want to know what's going on under the hood.

**Comment by AndrewBelt** [comment]
The engine buffer size is "1" if you could call it that. Rack doesn't know anything about audio. It just pushes the values down wires, steps modules, and repeats. Modules like Audio Interface can intercept those values and buffer them into audio drivers.

**Comment by squinkylabs** [comment]
ok, tx!
